src/link.cpp

```diff
@@ -1,21 +1,74 @@
 #include "link.h"
+typedef u8 uint8_t;
+#include <sstream>
 
 u8 FMS::curByte = 0;
 u8 FMS::curByte2 = 1;
 
-void FMS::startTransfer() {
+u8 secondShake[8] = { 0xA5, 0x00, 0x84, 0x01, 0x04, 0x04, 0x57, 0xd2 };
+u8 firstShake[8] = { 0xA5, 0x00, 0x84, 0x01, 0x03, 0x04, 0x70, 0x31 };
+
+//copied from keyboard example:
+static SwkbdState swkbd;
+static char mybuf[60];
+static SwkbdStatusData swkbdStatus;
+static SwkbdLearningData swkbdLearning;
+SwkbdButton button = SWKBD_BUTTON_NONE;
+bool didit = false;
+
+void FMS::startTransfer(int first = 0, bool editdata = false) {
     std::cout << ":: Start listening" << std::endl;
     
     const size_t SIZE = 0x1000;
     u8* data;
     data = (u8*) memalign(SIZE, SIZE);
+
+	int firstshakedone = 0;
     
-    u8 firstShake[8] = {0xA5, 0x00, 0x84, 0x01, 0x03, 0x04, 0x0A, 0x50};
-    u8 secondShake[8] = {0xA5, 0x00, 0x84, 0x01, 0x04, 0x04, 0x4C, 0x00};
-    
+	u8 dataSend[8];
+	std::copy(firstShake, firstShake + 8, dataSend);
 	Result ret = iruInit((u32*) data, SIZE);
     printIfError(ret);
 
+	if (editdata == true) {
+		cout << "note: this is a beta feature and will crash, press (X) to continue (Y) to go back";
+		while (1) {
+			if (hidKeysDown() & KEY_Y) {
+				return;
+			}
+			if (hidKeysDown() & KEY_X) {
+				break;
+			}
+		}
+		consoleClear();
+		std::string temp = FMS::u8tostring(firstShake, 8);
+		const char * c = temp.c_str();
+		std::cout << "original: " << temp << std::endl;
+		didit = true;
+		swkbdInit(&swkbd, SWKBD_TYPE_NORMAL, 3, -1);
```

Comment:
My guess why the keyboard probably doesn't work is because `aptMainLoop()` isn't called at all in this function, but that's my guess for now. In the future, this link will probably will be a separate thread from the main.

Reply:
the keyboard does work (i've tested it before), but it crashes when i try to convert the string back into a u8 array